File: clients/typescript/test/pkce.test.ts
import { describe, it, expect } from "vitest";
import { createPkceChallenge, randomState } from "../src/pkce.js";

describe("pkce", () => {
  it("produces a verifier and a base64url SHA-256 challenge", async () => {
    const { verifier, challenge } = await createPkceChallenge();
    expect(verifier).toMatch(/^[A-Za-z0-9\-._~]{43,128}$/);
    expect(challenge).toMatch(/^[A-Za-z0-9\-_]+$/); // base64url, no padding
    expect(challenge).not.toContain("=");
  });

  it("generates unique random state strings", () => {
    const a = randomState();
    const b = randomState();
    expect(a).not.toBe(b);
    expect(a.length).toBeGreaterThanOrEqual(16);
  });
});

File: clients/dart/test/pkce_test.dart
import 'package:test/test.dart';
import 'package:zigbase_client/zigbase_client.dart';

final _unreservedPattern = RegExp(r'^[A-Za-z0-9\-._~]+$');

void main() {
  group('createPkceChallenge', () {
    test('verifier is 64 chars from the unreserved charset', () {
      final pkce = createPkceChallenge();
      expect(pkce.verifier.length, 64);
      expect(_unreservedPattern.hasMatch(pkce.verifier), isTrue);
    });

    test('challenge is derived from the verifier', () {
      final pkce = createPkceChallenge();
      expect(pkce.challenge, pkceChallengeFor(pkce.verifier));
    });

    test('two calls produce different verifiers', () {
      final a = createPkceChallenge();
      final b = createPkceChallenge();
      expect(a.verifier, isNot(b.verifier));
    });
  });

  group('pkceChallengeFor', () {
    test('matches the RFC 7636 appendix B vector', () {
      const verifier = 'dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk';
      const expected = 'E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM';
      expect(pkceChallengeFor(verifier), expected);
    });
  });

  group('randomState', () {
    test('is 32 chars from the unreserved charset', () {
      final state = randomState();
      expect(state.length, 32);
      expect(_unreservedPattern.hasMatch(state), isTrue);
    });

    test('two calls differ', () {
      expect(randomState(), isNot(randomState()));
    });
  });
}
